File: java/src/com/liveperson/kafka/producer/BinaryProducer.java
package com.liveperson.kafka.producer;


import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;

import javax.xml.bind.DatatypeConverter;
import java.util.*;

/**
 * Created by elio on 1/5/17.
 */
public class BinaryProducer extends BaseProducer {


    public BinaryProducer(Properties props){
        super(props);
    }

    @Override
    protected KafkaProducer createProducer(Properties props) {
        return new KafkaProducer<byte[], byte[]>(props);
    }

    @Override
    public void sendWithKey(String msgId, String topic, String value, String key) {
        ProducerRecord<byte[], byte[]> producerRecord = new ProducerRecord<byte[], byte[]>(topic, DatatypeConverter.parseBase64Binary(value), DatatypeConverter.parseBase64Binary(key));
        send(msgId, producerRecord);
    }

    @Override
    public void sendWithKeyAndPartition(String msgId, String topic, String value, String key, Integer partition) {
        ProducerRecord<byte[], byte[]> producerRecord = new ProducerRecord<byte[], byte[]>(topic, partition, DatatypeConverter.parseBase64Binary(value), DatatypeConverter.parseBase64Binary(key));
        send(msgId, producerRecord);
    }

    @Override
    public void send(String msgId, String topic, String value) {
        ProducerRecord<byte[], byte[]> producerRecord = new ProducerRecord<byte[], byte[]>(topic, DatatypeConverter.parseBase64Binary(value));
        send(msgId, producerRecord);
    }
}

File: lib/binaryProducer.js
/**
 * Created by elio on 1/9/17.
 */
const Producer = require('./producer');

class BinaryProducer extends Producer{
    constructor(options) {
        super(options, "com.liveperson.kafka.producer.BinaryProducer", "org.apache.kafka.common.serialization.ByteArraySerializer");
    }

    send(topic, msg, cb) {
        super.send(topic, msg.toString('binary'), cb);
    }

    sendWithKey(topic, msg, key, cb) {
       super.sendWithKey(topic, msg.toString('binary'), key.toString('binary'), cb);
    }

    sendWithKeyAndPartition(topic, msg, key, partition, cb) {
        super.sendWithKeyAndPartition(topic, msg.toString('binary'), key.toString('binary'), partition, cb);
    }
}

module.exports = BinaryProducer;
